### ISSUE TITLE
libsamplerate_scale being reset due to locale bug

### ISSUE THREAD

**Issue by pwseo** [fabiangreffrath/crispy-doom#620]
### Background

Version of Crispy Doom: 5.6.4 (from the OS repo) and 5.8 (compiled from source)

Operating System and version: Kubuntu 20.04

Game: Doom II

Any loaded WADs and mods (please include full command line):
`crispy-doom -iwad ~/doom2.wad -deh ~/Downloads/btsx_e1/btsx_e1.deh -file ~/Downloads/btsx_e1/*.wad`

### Bug description
Whenever I play Back to Saturn X WAD, the value of `libsamplerate_scale` is rewritten as I quit Crispy Doom. As it is rewritten to `0,000000` (yes, with a comma), all SFX becomes silent in further `crispy-doom` executions.

Observed behavior:
If i start up Crispy with the command line above, scroll through the menu options and quit (even without starting a new game) and then grep `crispy-doom.cfg` for `libsamplerate_scale`, I get the following (I deleted crispy's output to make it more readable):
```
$ grep libsamplerate_scale crispy-doom.cfg
libsamplerate_scale           0.650000
$ crispy-doom [args like above] && grep libsamplerate_scale crispy-doom.cfg 
libsamplerate_scale           0,650000
$ crispy-doom [args like above] && grep libsamplerate_scale crispy-doom.cfg 
libsamplerate_scale           0,000000
```

Expected behavior:
The value shouldn't change.

Further notes:
Noticing the value was first changed from `0.650000` to `0,650000` and then to `0,000000` I thought it could be due to some locale mishap (I have a mix of en_US.UTF-8 and pt_PT.UTF-8 on my system because I like my OS to be in english, but need both numerics and currency in portuguese).
So I ran the test above but prepended `LC_ALL=C` to every `crispy-doom` command line: `libsamplerate_scale` never changed, and the decimal separator was always the dot.
Though this is a quick band-aid, I'm curious as to why this happens in the first place, but am not experienced enough to track it down.

So, I'm not really sure if this is a crispy-doom bug or a bug from the WAD itself, though it seems odd the WAD should be causing this, hence my posting here. Please let me know if this isn't the right place for this discussion.

**Comment by turol** [comment]
Parsing it:

https://github.com/fabiangreffrath/crispy-doom/blob/56c7decb9958734876d4377a15dc4c7ed47713ff/src/m_config.c#L2441

Saving it:
https://github.com/fabiangreffrath/crispy-doom/blob/56c7decb9958734876d4377a15dc4c7ed47713ff/src/m_config.c#L2375

`atof` at least depends on locale. Apparently `printf` does too. We should use something that doesn't. This is probably also an upstream bug.

**Comment by fabiangreffrath** [comment]
> This is probably also an upstream bug.

Most definitely.

**Comment by pwseo** [comment]
I thought both `atof()` and `fprintf()` needed a call to `setlocale()` to make the C runtime read the `LC_*` environment variables and use them. But indeed, I didn't find them while grepping the source.

**Comment by fabiangreffrath** [comment]
Actually, it appears to be the other way round. Those functions do always adhere to the locale settings, so if you want yo force them to use non-localized formats, you'll have to put these away with `setlocale()` before.

**Comment by fabiangreffrath** [comment]
The question is if we really want to support changing locales between to runs of Chocolate or Crispy Doom using the same config. Maybe @fragglet can shed some light upon this?

**Comment by pwseo** [comment]
> Those functions do always adhere to the locale settings

It's odd, in simple toy C applications my locale settings are always ignored; I have to call `setlocale()` to make the application honour them.

> The question is if we really want to support changing locales between to runs of Chocolate or Crispy Doom using the same config. Maybe @fragglet can shed some light upon this?

Not quite: I only experienced this bug with that specific WAD. Also, I didn't change locales between consecutive runs of Crispy Doom; everything happened with `LANG=en_US.UTF-8` and `LC_NUMERIC=pt_PT.UTF-8` **and** trying to play the Back to Saturn X WAD.
I only changed `LC_ALL` to `C` to see if the bug persisted across consecutive runs in a "less controversial" locale setting.

Consecutive runs of Crispy Doom with "classic" WADs never made this happen (with the same locale settings).

**Comment by fabiangreffrath** [comment]
That's really confusing! A PWAD should never have any effect on the format for floating point number input or output. 😕

**Comment by kitchen-ace** [comment]
Could it be something to do with both loading the wad (which has an embedded DEHACKED lump with BEX format strings) and the .deh patch (which is a vanilla compatible patch)? 

You shouldn't actually need to load the deh on the command-line with Crispy Doom.

**Comment by pwseo** [comment]
> That's really confusing! A PWAD should never have any effect on the format for floating point number input or output. confused

That was one of the reasons I filed the issue.

> Could it be something to do with both loading the wad (which has an embedded DEHACKED lump with BEX format strings) and the .deh patch (which is a vanilla compatible patch)?

It doesn't seem like it is. I just tried omitting the .deh patch and the behaviour is the same as described above.

**Comment by pwseo** [comment]
@fabiangreffrath, @fragglet I have some new data.

A call to `setlocale()` was indeed the culprit. I first pinpointed the locale change to the `I_SDL_RegisterSong()` function, and inside it, the `Mix_LoadMus()` function, which belongs to SDL. Then, I just used `gdb` to break execution at the next `setlocale()` call, which came from `libinstpatch` (used by SDL via fluidsynth).
When I set out to investigate, this was the first result from Google: issue swami/libinstpatch#37

I recompiled the library from the latest sources and everything seems to be working fine now.
